---
title: "RNAseq"
author: "Howard Cen"
date: "2024-11-16"
output: html_document
---

```{r}

library(tidyverse)
#BiocManager::install("DESeq2")
library("DESeq2")

BiocManager::install("apeglm")

#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)

#BiocManager::install("clusterProfiler")
#BiocManager::install("ReactomePA")
library("clusterProfiler")
library("ReactomePA")

#install.packages("devtools") # most recent version of complexheatmap
#library(devtools)
#install_github("jokergoo/ComplexHeatmap")
library(ComplexHeatmap)
library(circlize) 
#install.packages("gridtext")
library(gridtext)
library(scales)

#install.packages("ggrepel")
library(ggrepel) # https://ggrepel.slowkow.com/articles/examples.html

setwd(dirname(rstudioapi::getActiveDocumentContext()$path)) ##Set working directory to where this file is.
getwd()
```


```{r processing}

#files <- list.files(path="input", pattern = "genes\\.sf$") # List all files ending with genes.sf

df1 <- read.delim("input/D1-JZ-091124_711853222.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(DMSO_1 = NumReads)


df2 <- read.delim("input/D2-JZ-091124_711853225.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(DMSO_2 = NumReads)

df3 <- read.delim("input/D3-JZ-091124_711853224.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(DMSO_3 = NumReads)

df4 <- read.delim("input/X1-JZ-091124_711853255.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(ISX_1 = NumReads)

df5 <- read.delim("input/X2-JZ-091124_711853223.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(ISX_2 = NumReads)

df6 <- read.delim("input/X3-JZ-091124_711853252.quant.genes.sf") %>%
  mutate(ENSG = gsub("\\..*","", Name)) %>%
  select(all_of(c('Name', 'ENSG','NumReads'))) %>%
  rename(ISX_3 = NumReads)

df <- full_join(df1,df2) %>%
  full_join(df3)%>%
  full_join(df4)%>%
  full_join(df5)%>%
  full_join(df6)

View(df)

write.table(df, "output/rawcounts.txt", row.names = F)


```

```{r processing}
df <- read.table("output/rawcounts.txt", header = T)
dup <- df %>% group_by(ENSG) %>% filter(n()>1)
View(dup)
#aggr <- dup %>% group_by(ENSG) %>% summarise(across(everything(), max, na.rm = TRUE))

length(unique(dup$ENSG)) # 45 genes

df <- filter(df, !grepl("PAR_Y", Name))

raw <- df %>% column_to_rownames("ENSG") %>% select(-"Name")
View(raw)

raw.counts <- round(raw,0) # Counts are not integers. Rounded up because DESeq2 requires integers.
raw.counts <- raw.counts[rowSums(raw.counts)!=0,] # Counts<0.5 became 0. Remove genes with all 0 again
dim(raw.counts) # 32063 genes left
summary(colSums(raw.counts))
#    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#14990023 17462594 19239534 18960576 20746653 22175062 

raw.counts$ENSG <- rownames(raw.counts)
raw.counts$entrez <- mapIds(org.Hs.eg.db, keys=raw.counts$ENSG, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
raw.counts$symbol <- mapIds(org.Hs.eg.db, keys=raw.counts$ENSG, column="SYMBOL", keytype="ENSEMBL", multiVals="first")
raw.counts$fullname <- mapIds(org.Hs.eg.db, keys=raw.counts$ENSG, column="GENENAME", keytype="ENSEMBL", multiVals="first")

View(raw.counts)

write.table(raw.counts, "output/rawcounts_names.txt", row.names = F)
write.csv(raw.counts, "output/rawcounts_names.csv", row.names = F)

# meta data

sample <- colnames(raw.counts)
group <- gsub("_[0-9]","", sample) 
group <- gsub ("ISX", "ISX9", group)
meta.data <- data.frame(sample, group)
row.names(meta.data) <- meta.data$sample
all(colnames(raw.counts)==rownames(meta.data))
meta.data$group <- factor(meta.data$group, levels=c("DMSO",
                                                    "ISX9"))
View(meta.data)

```

```{r}
# DESeq2 and PCA =============

count.data.set <- DESeqDataSetFromMatrix(countData=raw.counts, 
                                         colData=meta.data, design= ~ group) 

# Filter out low count
nrow(count.data.set) # 32063 genes

keep <- rowSums(counts(count.data.set)>=5) >= ncol(raw.counts)*0.5  # Keep genes with >5 counts in at least 50% samples. 
count.filter <- count.data.set[keep,]
nrow(count.filter) # 20660 genes left.

# create DESeq object
count.data.set.object <- DESeq(count.filter)

# normalized counts (without VST)
dds <- estimateSizeFactors(count.data.set.object)
normalized.counts <- counts(dds, normalized=TRUE)

normalized.counts <- as.data.frame(normalized.counts)
is.data.frame(normalized.counts)
normalized.counts$ENSG <- rownames(normalized.counts)
normalized.counts$entrez <- mapIds(org.Hs.eg.db, keys=normalized.counts$ENSG, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
normalized.counts$symbol <- mapIds(org.Hs.eg.db, keys=normalized.counts$ENSG, column="SYMBOL", keytype="ENSEMBL", multiVals="first")
normalized.counts$fullname <- mapIds(org.Hs.eg.db, keys=normalized.counts$ENSG, column="GENENAME", keytype="ENSEMBL", multiVals="first")
View(normalized.counts)
write.csv(normalized.counts, file="output/norm_counts.csv", row.names = F)

# normalized  VST counts
vsd <- vst(count.data.set.object)
norm.data = assay(vsd)
norm.data <- as.data.frame(norm.data)
norm.data$ENSG <- rownames(norm.data)
norm.data$entrez <- mapIds(org.Hs.eg.db, keys=norm.data$ENSG, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
norm.data$symbol <- mapIds(org.Hs.eg.db, keys=norm.data$ENSG, column="SYMBOL", keytype="ENSEMBL", multiVals="first")
norm.data$fullname <- mapIds(org.Hs.eg.db, keys=norm.data$ENSG, column="GENENAME", keytype="ENSEMBL", multiVals="first")
View(norm.data)
write.csv(norm.data,file="output/norm_VST.csv", row.names = F)

# cluster
sampleDists <- dist(t(norm.data),  method = "euclidean") # "euclidean", "maximum", "manhattan", "canberra", "binary" or "minkowski"
clusters=hclust(sampleDists)
plot(clusters)

# PCA
# modify the PCA function to change format ---------------------------

getMethod("plotPCA","DESeqTransform")
plotPCA.format <- function (object, ...) 
{
  .local <- function (object, intgroup = "condition", 
                      ntop = 500, returnData = FALSE) 
  {
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                       length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
      stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
                                                 drop = FALSE])
    group <- if (length(intgroup) > 1) {
      factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
      colData(object)[[intgroup]]
    }
    d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], group = group, 
                    intgroup.df, name = colnames(object))
    if (returnData) {
      attr(d, "percentVar") <- percentVar[1:2]
      return(d)
    }
    ggplot(data = d, aes_string(x = "PC1", y = "PC2", 
                                fill = "group")) + geom_point(size = 4, shape=21,alpha=0.6) + 
      xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
      ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) + 
      coord_fixed() #+  geom_label_repel((aes(label=sample)))
  }
  .local(object, ...)
}


#------------------------------------------------------------------------

# use color blind friendly palette
cbPalette <- c("#E69F00", #lightorange
               "#56B4E9", #blue
               "#D55E00", #darkorange
               "#009E73", #green
               "#CC79A7", #magenta
               "#0072B2", #darkblue
               "#F0E442", #yellow
               "#999999" #grey
)


# plot PCA

p <- plotPCA.format(vsd, intgroup=c("group"))+ 
  geom_text_repel(aes(label=colData(vsd)$sample),size=3,
                  color="grey50",
                  box.padding   = 0.4,
                  point.padding = 0,
                  #force=1,
                  #force_pull=10,
                  max.overlaps = Inf, # always show all label, regardless of overlap
                  #min.segment.length = 0, # always draw line
                  segment.color = 'grey50')+
  scale_fill_manual(values=cbPalette) +
  theme_bw()+
  theme(panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x  = element_blank(),
        panel.grid.major.x  = element_blank(),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 16),
        legend.text = element_text(size = 16),
        legend.title = element_blank()
  )+
  theme(aspect.ratio=1/1)

p

# FYI, this is the default function plotPCA()
#plotPCA(vsd, intgroup=c("group"))

ggsave(plot = p, filename="figures/PCA_label.pdf",width=13,height=10,units="cm")
ggsave(plot = p, filename="figures/PCA_label.svg",device= svg,width=13,height=10,units="cm")

```

```{r}

# sample matrix

library("RColorBrewer")
sampleDistMatrix <- as.matrix( sampleDists )
View(sampleDistMatrix)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         legend = T,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

pDist<-pheatmap(sampleDistMatrix,
                legend = F,
                clustering_distance_rows = sampleDists,
                clustering_distance_cols = sampleDists,
                cellwidth = 20, cellheight = 20,
                treeheight_row=10,
                treeheight_col=10,
                col = colors)

pDist

# ran in console
{png(file = "figures/dist_matrix.png",
    width = 10, 
    height = 10,
    res  =  400,
    unit = "cm")

pDist

dev.off()}
```

png(file = "figures/dist_matrix.png",
      width = 10, 
      height = 10,
      res  =  400,
      unit = "cm")
     
pDist
     
dev.off()


```{r}
# plot the VST normalized counts on heatmap =============


View(norm.data)
m<- norm.data

#m <- as.matrix(as.data.frame(lapply(m.anno, as.numeric),check.names=F))
View(m)
m.z <- t(scale(t(m))) #%>% as.data.frame()
View(m.z)
colnames(m.z)
ceiling(max(abs(m.z)))
#m.z[m.z>5] <- NA
#m.z <- t(scale(t(m.z)))
#m.z[is.na(m.z)] <- 5

number_of_d <- length(grep("DMSO", colnames(m.z)))
number_of_x <- length(grep("ISX", colnames(m.z)))

end_index_d <- grep("DMSO", colnames(m.z))[number_of_d]
end_index_x <- grep("ISX", colnames(m.z))[number_of_x]

###

#heatmap_list_total <- 
heatmap.all <- Heatmap(m.z, #matrix_Z_score_total,
                         name = "Z score",
                         show_row_names = FALSE,
                         show_column_names = FALSE,
                         show_row_dend = TRUE,
                         # row_labels = gt_render(m.anno$protein.symbol),
                         row_names_gp = gpar(fontsize = 8),
                         column_names_side = "top",
                         column_dend_side = "bottom",
                     clustering_distance_rows = "euclidean",
                     clustering_method_rows = "ward.D2",
                     row_dend_side = "left",
                     row_dend_width = unit(5, "mm"),
                         layer_fun = function(j, i, x, y, width, height, fill) {
                           mat = restore_matrix(j, i, x, y)
                           ind = unique(c(mat[, c(end_index_d#, 
                                                  #end_index_bud
                                                  )]))
                           grid.rect(x = x[ind] + unit(0.5/ncol(m.z), "npc"), 
                                     y = y[ind], 
                                     width = unit(0.03, "inches"), 
                                     height = unit(1/nrow(m.z), "npc"),
                                     gp = gpar(col = "white")
                           )
                         },
                         col = colorRamp2(c(-3,0,3), c("blue", "white", "red")),
                         top_annotation = columnAnnotation(empty = anno_empty(border = FALSE
                                                                              , height = unit(12, "mm")
                         )),
                         
                         column_order = 1:ncol(m.z),
                         height = 
                           
                           
                           unit(80, "mm"), 
                         
                         width = ncol(m.z)*unit(6, "mm"),
                         border_gp = gpar(col = "black"),
                         show_heatmap_legend = TRUE,
                         heatmap_legend_param = list(
                           title = "Z-score",
                           title_position = "topleft",
                           legend_height = unit(4, "cm")))

draw(heatmap.all)
#

png(file = "figures/heatmap_all.png",
    width = 3, 
    height = 4, 
    units = "in", res = 600)

{pdf(file = "figures/heatmap_all.pdf",
    width = 3, 
    height = 4)

draw(heatmap.all)

#
seekViewport("annotation_empty_1")
loc1 = deviceLoc(x = unit(0, "npc"), y = unit(0, "npc"))
loc2 = deviceLoc(x = unit(1, "npc"), y = unit(1, "npc"))

####Condition labels

#Condition label 1


grid.rect(x = (loc2$x - loc1$x)*(end_index_d)/2/ncol(m.z),
          y = 0,
          width = (loc2$x - loc1$x)*(number_of_d)/ncol(m.z), 
          height = (loc2$y - loc1$y)/2,
          just = c("center", "bottom"),
          gp = gpar(fill = alpha(cbPalette[1], 0.5),
                    col = alpha(cbPalette[1], 0.5)
          )
)
grid.text(expression("DMSO"), 
          x = (loc2$x - loc1$x)*(end_index_d)/2/ncol(m.z),
          y = 0.25,
          just = c("center", "center"),
          gp = gpar(fontsize = 11,
                    col = "black"))

#Condition label 2

grid.rect(x = (loc2$x - loc1$x)*(end_index_d + 
                                   end_index_x)/2/ncol(m.z),
          y = 0,
          width = (loc2$x - loc1$x)*(number_of_x)/ncol(m.z), 
          height = (loc2$y - loc1$y)/2,
          just = c("center", "bottom"),
          gp = gpar(fill = alpha(cbPalette[2], 0.5),
                    col = alpha(cbPalette[2], 0.5)
          )
)
grid.text(expression("ISX9"), 
          x = (loc2$x - loc1$x)*(end_index_d + 
                                   end_index_x)/2/ncol(m.z),
          y = 0.25,
          just = c("center", "center"),
          gp = gpar(fontsize = 11))





###Top label gaps

#Vertical lines
grid.rect(x = end_index_d/ncol(m.z),
          y = 0,
          height = (loc2$y - loc1$y)/2,
          width = unit(0.03, "inches"),
          just = c("center", "bottom"),
          gp = gpar(fill = "white", 
                    col = "white", 
                    lwd = 1
          ))


dev.off()}
```


```{r}

# DEG ============

#resultsNames(count.data.set.object)


#res <- results(count.data.set.object, contrast=c("group",exp,ctrl),alpha=0.05)
res <- results(count.data.set.object, contrast=c("group","ISX9","DMSO"),alpha=0.05)

# save result summary
summary(res) 
out <- capture.output(summary(res))
cat("ISX9 vs DMSO", out, file="output/results_summary.txt", sep="\n", append=TRUE)

# save results - all genes
res <-  na.omit(res) # omit NA
res  <-  res[order(res$padj),] 
res <- as.data.frame(res)

res$ENSG <- rownames(res)
res$entrez <- mapIds(org.Hs.eg.db, keys=res$ENSG, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
res$symbol <- mapIds(org.Hs.eg.db, keys=res$ENSG, column="SYMBOL", keytype="ENSEMBL", multiVals="first")
res$fullname <- mapIds(org.Hs.eg.db, keys=res$ENSG, column="GENENAME", keytype="ENSEMBL", multiVals="first")

#columns(org.Hs.eg.db)

View(res)
write.table(res, sep="\t",file=paste0("output/ISX_DE_results.txt"), row.names=TRUE,col.names=NA,quote=FALSE)
write.csv(res, "output/ISX_DE_results.csv", row.names = F)

### with lfcshrinkage

resLFC <- lfcShrink(count.data.set.object, coef="group_ISX9_vs_DMSO", type="apeglm")
# using 'apeglm' for LFC shrinkage. Cite: Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for sequence count data: removing the noise and preserving large differences. Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895
resLFC

summary(resLFC) 
out <- capture.output(summary(resLFC))
cat("ISX9 vs DMSO lfc shrinkage", out, file="output/results_summary.txt", sep="\n", append=TRUE)

# save results - all genes
resLFC <-  na.omit(resLFC) # omit NA
resLFC  <-  resLFC[order(resLFC$padj),] 
resLFC <- as.data.frame(resLFC)

resLFC$ENSG <- rownames(resLFC)
resLFC$entrez <- mapIds(org.Hs.eg.db, keys=resLFC$ENSG, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
resLFC$symbol <- mapIds(org.Hs.eg.db, keys=resLFC$ENSG, column="SYMBOL", keytype="ENSEMBL", multiVals="first")
resLFC$fullname <- mapIds(org.Hs.eg.db, keys=resLFC$ENSG, column="GENENAME", keytype="ENSEMBL", multiVals="first")

#columns(org.Hs.eg.db)

View(resLFC)
write.table(resLFC, sep="\t",file=paste0("output/ISX_DE_LFCshrinkage_results.txt"), row.names=TRUE,col.names=NA,quote=FALSE)
write.csv(resLFC, "output/ISX_DE_LFCshrinkage_results.csv", row.names = F)
```





